<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to sitePath • sitePath</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to sitePath">
<meta property="og:description" content="sitePath">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sitePath</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/sitePath.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wuaipinglab/sitePath/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="sitePath_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to sitePath</h1>
                        <h4 class="author">Chengyang Ji</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wuaipinglab/sitePath/blob/master/vignettes/sitePath.Rmd"><code>vignettes/sitePath.Rmd</code></a></small>
      <div class="hidden name"><code>sitePath.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Detection of site fixation in molecular evolution.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>sitePath</code> package does hierarchical search for fixation events given multiple sequence alignment and phylogenetic tree. These fixation events can be specific to a phylogenetic lineages or shared by multiple lineages. This is achieved by three major steps:</p>
<ol style="list-style-type: decimal">
<li>Import tree and sequence alignment</li>
<li>Resolve phylogenetic lineages</li>
<li>Hierarchical search for lineage-dependent fixation events</li>
</ol>
</div>
<div id="import-data" class="section level1">
<h1 class="hasAnchor">
<a href="#import-data" class="anchor"></a>Import data</h1>
<p>There’re various R packages for parsing phylogenetic tree and multiple sequence alignment files. For now, <code>sitepath</code> accepts <code>phylo</code> object and <code>alignment</code> object. Functions from <code>ggtree</code> and <code>seqinr</code> are able to handle most file formats.</p>
<div id="parse-phylogenetic-tree" class="section level2">
<h2 class="hasAnchor">
<a href="#parse-phylogenetic-tree" class="anchor"></a>Parse phylogenetic tree</h2>
<p>The S3 phylo class is a common data structure for phylogenetic analysis in R. The CRAN package <a href="https://cran.r-project.org/web/packages/ape/index.html">ape</a> provides basic parsing function for reading tree files. The Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/ggtree.html">ggtree</a> provides more comprehensive parsing utilities.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wuaipinglab.github.io/sitePath/">sitePath</a></span><span class="op">)</span>

<span class="va">tree_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ZIKV.newick"</span>, package <span class="op">=</span> <span class="st">"sitePath"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="va">tree_file</span><span class="op">)</span></pre></div>
<p>It is highly recommended that the file stores a rooted tree as R would consider the tree is rooted by default and re-rooting the tree in R is difficult.</p>
</div>
<div id="parse-and-match-sequence-alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#parse-and-match-sequence-alignment" class="anchor"></a>Parse and match sequence alignment</h2>
<p>Most multiple sequence alignment format can be parsed by <a href="https://cran.r-project.org/web/packages/seqinr/index.html">seqinr</a>. sitePath has a wrapper function for parsing and adding the sequence alignment.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">alignment_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ZIKV.fasta"</span>, package <span class="op">=</span> <span class="st">"sitePath"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phyMSAmatched.html">addMSA</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">alignment_file</span>, <span class="st">"fasta"</span><span class="op">)</span></pre></div>
</div>
</div>
<div id="phylogenetic-lineages" class="section level1">
<h1 class="hasAnchor">
<a href="#phylogenetic-lineages" class="anchor"></a>Phylogenetic lineages</h1>
<p>The names in tree and aligment must be matched. We use a <strong>tip-to-root</strong> algorithm to trim tree leaves/tips to expose the major branches. Before finding putative phylogenetic lineages, there involves a few more steps to evalute the impact of threshold on result.</p>
<div id="the-impact-of-threshold-on-resolving-lineages" class="section level2">
<h2 class="hasAnchor">
<a href="#the-impact-of-threshold-on-resolving-lineages" class="anchor"></a>The impact of threshold on resolving lineages</h2>
<p>In the current version, the resolving function only takes sequence similarity as one single threshold. The impact of threshold depends on the tree topology hence there is no universal choice. The function <code>sneakPeak</code> samples thresholds and calculates the resulting number of paths. <em>The use of this function can be of great help in choosing the threshold.</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">preassessment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lineagePath.html">sneakPeek</a></span><span class="op">(</span><span class="va">tree</span>, makePlot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/sneakPeek_plot-1.png" width="700"></p>
</div>
<div id="choose-a-threshold-by-yourself" class="section level2">
<h2 class="hasAnchor">
<a href="#choose-a-threshold-by-yourself" class="anchor"></a>Choose a threshold by yourself</h2>
<p>Use the return of the function <code>lineagePath</code> for downstream analysis. The choice of the threshold really depends. You can use the result from <code>sneakPeak</code> as a reference for threhold choosing. Here the default value is used.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lineagePath.html">lineagePath</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>
<span class="va">paths</span>
<span class="co">#&gt; This is a 'lineagePath' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 4 lineage paths using 5% as "major SNP" threshold</span></pre></div>
<p>You can visualize the result.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_paths-1.png" width="700"></p>
</div>
</div>
<div id="detecting-fixation-and-parallel-mutations" class="section level1">
<h1 class="hasAnchor">
<a href="#detecting-fixation-and-parallel-mutations" class="anchor"></a>Detecting fixation and parallel mutations</h1>
<p>Now you’re ready to find fixation and parallel mutations.</p>
<div id="entropy-minimization" class="section level2">
<h2 class="hasAnchor">
<a href="#entropy-minimization" class="anchor"></a>Entropy minimization</h2>
<p>The <code>sitesMinEntropy</code> function perform entropy minimization on every site for each lineage. The fixation and parallel mutations can be derived from the function’s return value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">minEntropy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sitesMinEntropy.html">sitesMinEntropy</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></pre></div>
</div>
<div id="fixation-mutations" class="section level2">
<h2 class="hasAnchor">
<a href="#fixation-mutations" class="anchor"></a>Fixation mutations</h2>
<p>The hierarchical search is done by <code>fixationSites</code> function. The function detects the site with fixation mutation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">fixations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixationSites.html">fixationSites</a></span><span class="op">(</span><span class="va">minEntropy</span><span class="op">)</span>
<span class="va">fixations</span>
<span class="co">#&gt; This is a 'fixationSites' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Result for 4 paths:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 988 139 894 2074 2086 2634 3045 107 1118 3353 1143 2842 3398 </span>
<span class="co">#&gt; No reference sequence specified. Using alignment numbering</span></pre></div>
<p>If you want to retrieve the result of a single site, you can pass the result of <code>fixationSites</code> and the site index to <code>extractSite</code> function. The output is a <code>sitePath</code> object which stores the tip names.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractSite.html">extractSite</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span></pre></div>
<p>It is also possible to retrieve the tips involved in the fixation of the site.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="../reference/extractTips.html">extractTips</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt;  [1] "ANK57896" "AMD61711" "AQS26698" "APG56458" "AUI42289" "AMR39834"</span>
<span class="co">#&gt;  [7] "AWH65848" "APO08504" "AMX81917" "AVZ47169" "AMX81916" "AMD61710"</span>
<span class="co">#&gt; [13] "AMK49492" "AMX81915" "AOC50652" "APH11611" "BBC70847" "AUF35022"</span>
<span class="co">#&gt; [19] "ATL14618" "AUF35021" "AVV62004" "BAX00477"</span>
<span class="co">#&gt; attr(,"AA")</span>
<span class="co">#&gt; [1] "S"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt;   [1] "BAV89190"     "AOI20067"     "AMM43325"     "AMM43326"     "AUI42329"    </span>
<span class="co">#&gt;   [6] "AUI42330"     "ANC90425"     "AMT75536"     "ANF16414"     "AMR68932"    </span>
<span class="co">#&gt;  [11] "ANA12599"     "AMM39806"     "AMR39830"     "AMV49165"     "AMO03410"    </span>
<span class="co">#&gt;  [16] "ANO46307"     "AVG19275"     "ANN44857"     "ANO46306"     "ANO46309"    </span>
<span class="co">#&gt;  [21] "ANO46305"     "ANO46303"     "ARB08102"     "ANO46302"     "AHZ13508"    </span>
<span class="co">#&gt;  [26] "ANO46304"     "ANO46301"     "ANO46308"     "AOG18296"     "AOO19564"    </span>
<span class="co">#&gt;  [31] "AUI42194"     "APC60215"     "AMQ48986"     "ATG29307"     "ART29828"    </span>
<span class="co">#&gt;  [36] "AWF93617"     "ATG29284"     "ATG29287"     "ATG29303"     "AWF93619"    </span>
<span class="co">#&gt;  [41] "AWF93618"     "AQM74762"     "AUD54964"     "AQM74761"     "ATG29306"    </span>
<span class="co">#&gt;  [46] "ASL68974"     "ATG29267"     "ASL68978"     "AQX32985"     "ATG29315"    </span>
<span class="co">#&gt;  [51] "AQZ41956"     "ARI68105"     "ASU55505"     "AQZ41949"     "ASL68979"    </span>
<span class="co">#&gt;  [56] "ATG29299"     "ATI21641"     "ATG29270"     "ATG29291"     "AOY08536"    </span>
<span class="co">#&gt;  [61] "ANO46297"     "ANO46298"     "AQZ41950"     "AQZ41951"     "ARU07183"    </span>
<span class="co">#&gt;  [66] "ANG09399"     "AQZ41954"     "AOY08533"     "AQZ41947"     "AQZ41948"    </span>
<span class="co">#&gt;  [71] "ATG29292"     "ATG29295"     "AOW32303"     "AVZ25033"     "AOC50654"    </span>
<span class="co">#&gt;  [76] "AQZ41953"     "ATG29301"     "ATG29276"     "APO08503"     "AMC13913"    </span>
<span class="co">#&gt;  [81] "AMC13912"     "APO39243"     "APO39229"     "AQZ41952"     "AQZ41955"    </span>
<span class="co">#&gt;  [86] "AMK49165"     "ARB07976"     "APB03018"     "AMC13911"     "APB03019"    </span>
<span class="co">#&gt;  [91] "ASU55416"     "ANK57897"     "AWH65849"     "AMZ03556"     "ASU55417"    </span>
<span class="co">#&gt;  [96] "ANW07476"     "APY24199"     "AMA12086"     "AMH87239"     "APY24198"    </span>
<span class="co">#&gt; [101] "APO36913"     "ALX35659"     "AOG18295"     "ANQ92019"     "AML81028"    </span>
<span class="co">#&gt; [106] "APY24200"     "AMD16557"     "ARU07074"     "AOX49264"     "AOX49265"    </span>
<span class="co">#&gt; [111] "AOY08518"     "ARB07962"     "AMX81919"     "AMM39805"     "ARX97119"    </span>
<span class="co">#&gt; [116] "AMB37295"     "AMK79468"     "AML82110"     "AMR39831"     "AMX81918"    </span>
<span class="co">#&gt; [121] "ANC90426"     "ALU33341"     "ASB32509"     "AMA12085"     "AMU04506"    </span>
<span class="co">#&gt; [126] "AMA12087"     "AMA12084"     "AQU12485"     "AMS00611"     "AMQ48981"    </span>
<span class="co">#&gt; [131] "AOY08538"     "APH11492"     "AOY08517"     "AOY08541"     "AOO54270"    </span>
<span class="co">#&gt; [136] "AND01116"     "ARB07967"     "ANF04752"     "AOE22997"     "APQ41782"    </span>
<span class="co">#&gt; [141] "APQ41786"     "ASU55393"     "ASU55404"     "ASU55423"     "ANB66182"    </span>
<span class="co">#&gt; [146] "ASU55425"     "ASU55420"     "AQX32986"     "ASU55422"     "APQ41784"    </span>
<span class="co">#&gt; [151] "ANC90428"     "ASU55415"     "ASU55418"     "ARM59239"     "ASU55408"    </span>
<span class="co">#&gt; [156] "ASU55424"     "ASU55390"     "ASU55419"     "ASU55391"     "AMM39804"    </span>
<span class="co">#&gt; [161] "ASU55411"     "ANB66183"     "ASU55421"     "AMZ03557"     "ASU55392"    </span>
<span class="co">#&gt; [166] "AQX32987"     "ASU55403"     "ASU55399"     "APQ41783"     "ANS60026"    </span>
<span class="co">#&gt; [171] "ANB66184"     "ASU55426"     "ASU55412"     "ASU55413"     "ASU55410"    </span>
<span class="co">#&gt; [176] "ASU55397"     "ASU55400"     "ASU55409"     "APB03017"     "ASU55395"    </span>
<span class="co">#&gt; [181] "ASU55396"     "AOY08524"     "ASU55394"     "ASU55414"     "ASU55405"    </span>
<span class="co">#&gt; [186] "AMC33116"     "ASU55406"     "ASU55398"     "ASU55407"     "AMQ34003"    </span>
<span class="co">#&gt; [191] "AMQ34004"     "ASU55401"     "ASU55402"     "ARU07076"     "AMK49164"    </span>
<span class="co">#&gt; [196] "APG56457"     "AOR82892"     "ATB53752"     "ANH10698"     "AOR82893"    </span>
<span class="co">#&gt; [201] "ARU07075"     "AMB18850"     "YP_009428568" "AMQ48982"     "ART29823"    </span>
<span class="co">#&gt; [206] "APW84876"     "ASK51714"     "ARB07953"     "APW84872"     "AOY08525"    </span>
<span class="co">#&gt; [211] "APW84873"     "AOY08535"     "AVZ25035"     "ARB07932"     "AOY08523"    </span>
<span class="co">#&gt; [216] "AOY08542"     "ASW34087"     "AOY08537"     "APB03020"     "ART29826"    </span>
<span class="co">#&gt; [221] "ART29825"     "AOS90220"     "AMN14620"     "APW84874"     "APW84875"    </span>
<span class="co">#&gt; [226] "BAV82373"     "AOS90221"     "AOS90224"     "APB03021"     "APO39232"    </span>
<span class="co">#&gt; [231] "AOS90223"     "APO39237"     "ANH22038"     "APW84877"     "APO39236"    </span>
<span class="co">#&gt; [236] "AOY08546"     "AOY08516"     "APO39233"     "AOS90222"     "AOO53981"    </span>
<span class="co">#&gt; [241] "AOY08521"     "AOO85388"     "APO39228"    </span>
<span class="co">#&gt; attr(,"AA")</span>
<span class="co">#&gt; [1] "N"</span></pre></div>
<p>Use <code>plot</code> on a <code>sitePath</code> object to visualize the fixation mutation of a single site. Alternatively, use <code>plotSingleSite</code> on an <code>fixationSites</code> object with the site specified.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_sitePath-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">fixations</span>, <span class="fl">139</span><span class="op">)</span>
<span class="co">#&gt; Warning: `mutate_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `mutate()` instead.</span>
<span class="co">#&gt; See vignette('programming') for more help</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_sitePath-2.png" width="700"></p>
<p>To have an overall view of the transition of fixation mutation, use <code>plot</code> on an <code>fixationSites</code> object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">fixations</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_fixations-1.png" width="700"></p>
</div>
<div id="parallel-mutation" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-mutation" class="anchor"></a>Parallel mutation</h2>
<p>Parallel mutation can be found by the <code>parallelSites</code> function. There are four ways of defining the parallel mutation: <code>all</code>, <code>exact</code>, <code>pre</code> and <code>post</code>. Here <code>exact</code> is used as an example.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">paraSites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parallelSites.html">parallelSites</a></span><span class="op">(</span><span class="va">minEntropy</span>, mutMode <span class="op">=</span> <span class="st">"exact"</span><span class="op">)</span>
<span class="va">paraSites</span>
<span class="co">#&gt; This is a 'parallelSites' object.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Result for 4 paths:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 105 1717 988 1264 1226 3076 106 573 2611 2787 2749 3162 1857 1016 1171 2634 2357 1327 3046 3328 940 1404 1180 </span>
<span class="co">#&gt; No reference sequence specified. Using alignment numbering</span></pre></div>
<p>The result of a single site can be visualized by <code>plotSingleSite</code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paraSites</span>, <span class="fl">105</span><span class="op">)</span>
<span class="co">#&gt; Warning: `filter_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `filter()` instead.</span>
<span class="co">#&gt; See vignette('programming') for more help</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span></pre></div>
<p><img src="sitePath_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
</div>
<div id="additional-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-functions" class="anchor"></a>Additional functions</h1>
<p>This part is extra and experimental but might be useful when pre-assessing your data. We’ll use an example to demonstrate.</p>
<div id="inspect-one-site" class="section level2">
<h2 class="hasAnchor">
<a href="#inspect-one-site" class="anchor"></a>Inspect one site</h2>
<p>The <code>plotSingleSite</code> function will color the tree according to amino acids if you use the output of <code>lineagePath</code> function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="fl">139</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_sites-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="fl">763</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/plot_sites-2.png" width="700"></p>
</div>
<div id="snp-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#snp-sites" class="anchor"></a>SNP sites</h2>
<p>An SNP site could potentially undergo fixation event. The <code>SNPsites</code> function predicts possible SNP sites and the result could be what you’ll expect to be fixation mutation. Also, a tree plot with mutation could be visualized with <code>plotMutSites</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SNPsites.html">SNPsites</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotMutSites.html">plotMutSites</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="va">snps</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-2.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu"><a href="../reference/plotSingleSite.html">plotSingleSite</a></span><span class="op">(</span><span class="va">paths</span>, <span class="va">snps</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="sitePath_files/figure-html/find_SNP-3.png" width="700"></p>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] sitePath_1.5.24  BiocStyle_2.17.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] treeio_1.13.1       tidyselect_1.1.0    xfun_0.18          </span>
<span class="co">#&gt;  [4] purrr_0.3.4         lattice_0.20-41     colorspace_1.4-1   </span>
<span class="co">#&gt;  [7] vctrs_0.3.4         generics_0.0.2      htmltools_0.5.0    </span>
<span class="co">#&gt; [10] yaml_2.2.1          rlang_0.4.8         pkgdown_1.6.1.9000 </span>
<span class="co">#&gt; [13] pillar_1.4.6        glue_1.4.2          RColorBrewer_1.1-2 </span>
<span class="co">#&gt; [16] rvcheck_0.1.8       lifecycle_0.2.0     stringr_1.4.0      </span>
<span class="co">#&gt; [19] munsell_0.5.0       gtable_0.3.0        ragg_0.3.1         </span>
<span class="co">#&gt; [22] memoise_1.1.0       evaluate_0.14       labeling_0.3       </span>
<span class="co">#&gt; [25] knitr_1.30          parallel_4.0.2      Rcpp_1.0.5         </span>
<span class="co">#&gt; [28] scales_1.1.1        backports_1.1.10    BiocManager_1.30.10</span>
<span class="co">#&gt; [31] desc_1.2.0          jsonlite_1.7.1      farver_2.0.3       </span>
<span class="co">#&gt; [34] systemfonts_0.3.2   fs_1.5.0            gridExtra_2.3      </span>
<span class="co">#&gt; [37] ggplot2_3.3.2       aplot_0.0.6         digest_0.6.25      </span>
<span class="co">#&gt; [40] stringi_1.5.3       bookdown_0.20       dplyr_1.0.2        </span>
<span class="co">#&gt; [43] ggrepel_0.8.2       ade4_1.7-15         grid_4.0.2         </span>
<span class="co">#&gt; [46] rprojroot_1.3-2     ggtree_2.3.6        tools_4.0.2        </span>
<span class="co">#&gt; [49] magrittr_1.5        lazyeval_0.2.2      patchwork_1.0.1    </span>
<span class="co">#&gt; [52] tibble_3.0.3        seqinr_3.6-1        crayon_1.3.4       </span>
<span class="co">#&gt; [55] ape_5.4-1           tidyr_1.1.2         pkgconfig_2.0.3    </span>
<span class="co">#&gt; [58] MASS_7.3-51.6       ellipsis_0.3.1      tidytree_0.3.3     </span>
<span class="co">#&gt; [61] assertthat_0.2.1    rmarkdown_2.4       R6_2.4.1           </span>
<span class="co">#&gt; [64] nlme_3.1-148        compiler_4.0.2</span></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chengyang Ji.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
